internal/v3/controller/connectionsecret/connectionsecret_controller.go

```diff
@@ -0,0 +1,356 @@
+		Watches(
+			&akov2.AtlasDeployment{},
+			handler.EnqueueRequestsFromMapFunc(r.newConnectionTargetMapFunc),
+			builder.WithPredicates(predicate.Or(
+				watch.ReadyTransitionPredicate(func(d *akov2.AtlasDeployment) bool {
+					return api.HasReadyCondition(d.Status.Conditions)
+				}),
+				predicate.GenerationChangedPredicate{},
+			)),
+		).
+		Watches(
+			&akov2.AtlasDataFederation{},
+			handler.EnqueueRequestsFromMapFunc(r.newConnectionTargetMapFunc),
+			builder.WithPredicates(predicate.Or(
+				watch.ReadyTransitionPredicate(func(d *akov2.AtlasDataFederation) bool {
+					return api.HasReadyCondition(d.Status.Conditions)
+				}),
+				predicate.GenerationChangedPredicate{},
+			)),
+		).
+		WithOptions(controller.TypedOptions[reconcile.Request]{
+			RateLimiter:        ratelimit.NewRateLimiter[reconcile.Request](),
+			SkipNameValidation: pointer.MakePtr(skipNameValidation),
+		}).
+		Complete(r)
+}
+
+func allowsByScopes(u *akov2.AtlasDatabaseUser, epName string, epType akov2.ScopeType) bool {
+	scopes := u.Spec.Scopes
+	filtered_scopes := u.GetScopes(epType)
+	if len(scopes) == 0 || stringutil.Contains(filtered_scopes, epName) {
+		return true
+	}
+
+	return false
+}
+
+func (r *ConnectionSecretReconciler) generateConnectionSecretRequests(users []akov2.AtlasDatabaseUser) []reconcile.Request {
+	reqs := make([]reconcile.Request, 0, len(users))
+	for _, u := range users {
+		reqs = append(reqs, reconcile.Request{
+			NamespacedName: types.NamespacedName{Namespace: u.Namespace, Name: u.Name},
+		})
+	}
+	return reqs
+}
+
+// listConnectionTargetsByProject retrieves all of the connectionTargets that live under an AtlasProject
+func (r *ConnectionSecretReconciler) listConnectionTargetsByProject(ctx context.Context, projectID string) ([]ConnectionTarget, error) {
+	var out []ConnectionTarget
+
+	for _, kind := range r.ConnectionTargetKinds {
+		switch kind.(type) {
+		case DataFederationConnectionTarget:
+			list := &akov2.AtlasDataFederationList{}
+			if err := r.Client.List(ctx, list, &client.ListOptions{
+				FieldSelector: kind.SelectorByProjectID(projectID),
+			}); err != nil {
+				return nil, err
+			}
+
+			for i := range list.Items {
+				out = append(out, DataFederationConnectionTarget{
+					obj:             &list.Items[i],
+					client:          r.Client,
+					provider:        r.AtlasProvider,
+					globalSecretRef: r.GlobalSecretRef,
+					log:             r.Log,
+				})
+			}
+
+		case DeploymentConnectionTarget:
+			list := &akov2.AtlasDeploymentList{}
+			if err := r.Client.List(ctx, list, &client.ListOptions{
+				FieldSelector: kind.SelectorByProjectID(projectID),
+			}); err != nil {
+				return nil, err
+			}
+
+			for i := range list.Items {
+				out = append(out, DeploymentConnectionTarget{
+					obj:             &list.Items[i],
+					client:          r.Client,
+					provider:        r.AtlasProvider,
+					globalSecretRef: r.GlobalSecretRef,
+					log:             r.Log,
+				})
+			}
+		}
+	}
```

Comment:
Can we make it simpler and avoid code repetition?

```suggestion
	for _, kind := range r.ConnectionTargetKinds {
		list := ListForKind(kind) // this method has the switch case instead
		if err := r.Client.List(ctx, list, &client.ListOptions{
			FieldSelector: kind.SelectorByProjectID(projectID),
		}); err != nil {
			return nil, err
		}

		for i := range list.Items {
			out = append(out, DataFederationConnectionTarget{
				obj:             &list.Items[i],
				client:          r.Client,
				provider:        r.AtlasProvider,
				globalSecretRef: r.GlobalSecretRef,
				log:             r.Log,
			})
		}
	}
```